lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,859 @@
+  end
+
+  defp validate_and_start_child(supervisor, child_spec) do
+    case validate_child(child_spec) do
+      {:ok, child} -> call(supervisor, {:start_child, child})
+      error -> {:error, error}
+    end
+  end
+
+  defp validate_child(%{id: _, start: {mod, _, _} = start} = child) do
+    restart = Map.get(child, :restart, :permanent)
+    type = Map.get(child, :type, :worker)
+    modules = Map.get(child, :modules, [mod])
+
+    shutdown =
+      case type do
+        :worker -> Map.get(child, :shutdown, 5_000)
+        :supervisor -> Map.get(child, :shutdown, :infinity)
+      end
+
+    validate_child(start, restart, shutdown, type, modules)
+  end
+
+  defp validate_child({_, start, restart, shutdown, type, modules}) do
+    validate_child(start, restart, shutdown, type, modules)
+  end
+
+  defp validate_child(other) do
+    {:invalid_child_spec, other}
+  end
+
+  defp validate_child(start, restart, shutdown, type, modules) do
+    with :ok <- validate_start(start),
+         :ok <- validate_restart(restart),
+         :ok <- validate_shutdown(shutdown),
+         :ok <- validate_type(type),
+         :ok <- validate_modules(modules) do
+      {:ok, {start, restart, shutdown, type, modules}}
+    end
+  end
+
+  defp validate_start({m, f, args}) when is_atom(m) and is_atom(f) and is_list(args), do: :ok
+  defp validate_start(mfa), do: {:invalid_mfa, mfa}
+
+  defp validate_type(type) when type in [:supervisor, :worker], do: :ok
+  defp validate_type(type), do: {:invalid_child_type, type}
+
+  defp validate_restart(restart) when restart in [:permanent, :temporary, :transient], do: :ok
+  defp validate_restart(restart), do: {:invalid_restart_type, restart}
+
+  defp validate_shutdown(shutdown) when is_integer(shutdown) and shutdown > 0, do: :ok
+  defp validate_shutdown(shutdown) when shutdown in [:infinity, :brutal_kill], do: :ok
+  defp validate_shutdown(shutdown), do: {:invalid_shutdown, shutdown}
+
+  defp validate_modules(:dynamic), do: :ok
+
+  defp validate_modules(mods) do
+    if is_list(mods) and Enum.all?(mods, &is_atom/1) do
+      :ok
+    else
+      {:invalid_modules, mods}
+    end
+  end
+
+  @doc """
+  Terminates the given child identified by child id.
+
+  If successful, this function returns `:ok`. If there is no process with
+  the given PID, this function returns `{:error, :not_found}`.
+  """
+  @spec terminate_child(Supervisor.supervisor(), pid) :: :ok | {:error, :not_found}
+  def terminate_child(supervisor, pid) when is_pid(pid) do
+    call(supervisor, {:terminate_child, pid})
+  end
+
+  @doc """
+  Returns a list with information about all children.
+
+  Note that calling this function when supervising a large number
+  of children under low memory conditions can cause an out of memory
+  exception.
+
+  This function returns a list of tuples containing:
+
+    * `id` - as defined in the child specification but is always
+      set to `:undefined` for dynamic supervisors
```

Comment:
This looks weird. This is a dynamic supervisor, so this value will always be `:undefined`.